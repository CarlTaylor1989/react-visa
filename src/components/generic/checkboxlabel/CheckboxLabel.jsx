import React from 'react';
import PropTypes from 'prop-types';

import './CheckboxLabel.scss';

const CheckboxLabel = ({
  disabled,
  elementId,
  label,
  isSelected,
  onCheckboxChange
}) => (
  <div className={`checkboxLabel${disabled ? ' disabled' : ''}`}>
    <label htmlFor={elementId}>
      <input
        id={elementId}
        type="checkbox"
        disabled={disabled}
        checked={isSelected}
        onChange={() => onCheckboxChange(elementId)}
        className="checkbox"
      />
      <span className="label">{label}</span>
      <span className="checkmark" />
    </label>
  </div>
);

CheckboxLabel.defaultProps = {
  disabled: false
};

CheckboxLabel.propTypes = {
  disabled: PropTypes.bool,
  elementId: PropTypes.string.isRequired,
  isSelected: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  onCheckboxChange: PropTypes.func.isRequired
};

export default CheckboxLabel;
