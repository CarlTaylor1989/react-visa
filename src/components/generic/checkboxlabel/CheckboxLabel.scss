@import '../../../styles/mixins';
@import '../../../styles/variables';

$checkbox-width: 32px;
$checkbox-height: 32px;
$checkmark-width: 24px;
$checkmark-height: 24px;
$tick-icon: '../../../styles/icons/clients/tick_checkbox.svg';
$tick-icon-light: '../../../styles/icons/clients/tick_checkbox_white.svg';
$tick-icon-width: 18px;
$tick-icon-height: 14px;

.checkboxLabel {
  @include user-select;
  font-size: 18px;
  padding: 10px 0;
  position: relative;
  text-align: left;

  label {
    @include align-items(center);
    cursor: pointer;
    display: flex;
    justify-content: space-between;
  }

  input {
    @include opacity(0);
    height: 0;
    position: absolute;
    width: 0;

    &:checked ~ .checkmark {
      &::before,
      &::after {
        @include opacity(1);
      }
    }

    &:disabled ~ .checkmark {
      border-color: transparent;

      &::before {
        background-color: $colour-button;
      }

      &::after {
        background-image: url($tick-icon-light);
      }
    }

    &:disabled {
      &:checked ~ .checkmark {
        &::before,
        &::after {
          @include opacity(.6);
        }
      }

      &:not(:checked) + .label {
        @include opacity(.4);
      }
    }
  }

  .checkmark {
    @include border-radius(8px);
    border: 2px solid $colour-white;
    display: block;
    flex: 0 0 auto;
    height: $checkbox-height;
    margin-left: 14px;
    position: relative;
    width: $checkbox-width;

    &::before,
    &::after {
      @include opacity(0);
      content: '';
      display: block;
      left: 50%;
      position: absolute;
      top: 50%;
      transition: opacity .5s ease-in-out;
    }

    &::before {
      @include border-radius(3px);
      @include box-shadow(0, 0, 6px, 0, $colour-white);
      background-color: $colour-yellow;
      height: $checkmark-height;
      margin-left: -$checkmark-width / 2;
      margin-top: -$checkmark-width / 2;
      width: $checkmark-width;
    }

    &::after {
      background-image: url($tick-icon);
      height: $tick-icon-height;
      margin-left: -$tick-icon-width / 2;
      margin-top: -$tick-icon-height / 2;
      width: $tick-icon-width;
    }
  }

  &.disabled {
    label {
      cursor: not-allowed;
    }
  }

  &:hover {
    input:not(:disabled) ~ .checkmark { // sass-lint:disable-line force-pseudo-nesting
      &::before {
        @include opacity(.4);
      }
    }

    input:checked:not(:disabled) ~ .checkmark { // sass-lint:disable-line force-pseudo-nesting
      &::after {
        @include opacity(.4);
      }
    }
  }
}
