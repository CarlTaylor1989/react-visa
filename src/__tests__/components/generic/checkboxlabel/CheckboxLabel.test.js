/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import CheckboxLabel from '../../../../components/generic/checkboxlabel/CheckboxLabel';

let mockProps;
beforeEach(() => {
  mockProps = {
    disabled: false,
    elementId: 'irrelevant',
    isSelected: false,
    label: 'irrelevant',
    onCheckboxChange: jest.fn()
  };
});

test('renders the CheckboxLabel component with the enabled state', () => {
  const wrapper = shallow(<CheckboxLabel {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders the CheckboxLabel component with the disabled state', () => {
  const wrapper = shallow(<CheckboxLabel {...mockProps} disabled />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('calls the change event on the checkbox', () => {
  const wrapper = shallow(<CheckboxLabel {...mockProps} />);
  wrapper.find('input').at(0).simulate('change');
  expect(mockProps.onCheckboxChange).toHaveBeenCalledWith(mockProps.elementId);
});
