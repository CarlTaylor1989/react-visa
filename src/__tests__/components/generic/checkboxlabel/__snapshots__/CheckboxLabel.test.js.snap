// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the CheckboxLabel component with the disabled state 1`] = `
<div
  className="checkboxLabel disabled"
>
  <label
    htmlFor="irrelevant"
  >
    <input
      checked={false}
      className="checkbox"
      disabled={true}
      id="irrelevant"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="label"
    >
      irrelevant
    </span>
    <span
      className="checkmark"
    />
  </label>
</div>
`;

exports[`renders the CheckboxLabel component with the enabled state 1`] = `
<div
  className="checkboxLabel"
>
  <label
    htmlFor="irrelevant"
  >
    <input
      checked={false}
      className="checkbox"
      disabled={false}
      id="irrelevant"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="label"
    >
      irrelevant
    </span>
    <span
      className="checkmark"
    />
  </label>
</div>
`;
